Contour 2D node and Sort By Connections

vicdoval opened the pull request:
This heavy node creates 2D iso-contours from points and edges.
It has to be feeded width points with the same Z coordinate
It can work with constant values:
![isolines3](https://user-images.githubusercontent.com/10011941/34541729-ac2eb94e-f0d9-11e7-81bc-5b6c52f8dbe9.png)
Or weighted distances:
![isolines4](https://user-images.githubusercontent.com/10011941/34541756-c87e9628-f0d9-11e7-953f-f58c939d43a0.png)

Is focused on creating a valid contour, to do so the algorithm contains the "Intersect Edges" node, with its behaviour, (also the "mask vertices" node, and the "Remove doubles")
It is pretty experimental but works on most of the cases
I also added the listMatcher function to switch between long_cycle and long_repeat (it may be a good practice for other nodes)

To sort the contour I created the function SortbyConnexions and i added it to the "Vector Sort" node:

![vector_sort_by_connexion](https://user-images.githubusercontent.com/10011941/34542083-5200df4a-f0db-11e7-882f-83d5d22af691.png)

I hope it is useful and enjoyable

zeffii commented on nodes/modifier_make/isoLine2D.py:
```diff
@@ -0,0 +1,694 @@
+from sverchok.utils.sv_bmesh_utils import bmesh_from_pydata
+
+modeItems = [("Constant", "Constant", ""), ("Weighted", "Weighted", "")]
+listMatchItems = [("Long Repeat", "Long Repeat", ""), ("Long Cycle", "Long Cycle", "")]
```
add a space between this line and the new unrelated function below it.

zeffii commented on nodes/modifier_make/isoLine2D.py:
```diff
@@ -0,0 +1,694 @@
+    maskT = bpy.props.FloatProperty(
+        description = "Mask tolerance",
+        min=-1.0, default=1.0e-5, step=0.02, update=updateNode)
+    rad_ = FloatProperty(name='Radius', description='Radius',
```
sort out the sudden weird formatting here.  dont have spaces between varname=varvalue inside a function call.

zeffii commented on nodes/modifier_make/isoLine2D.py:
```diff
@@ -0,0 +1,694 @@
+        
+        return [listPlg]
+
+    def process(self):
```
clean up the whitespace in this function especially :)

zeffii replied:
i'd like to suggest using a pep8 checker or pylinter

zeffii replied:
to get into the habbit of formatting code in a way that is generally easier to read (because it will match the general style of python code found in this repo )

zeffii commented on nodes/modifier_make/isoLine2D.py:
```diff
@@ -0,0 +1,694 @@
+
+    # if this distance is larger than than VTX_PRECISION, we can skip it.
+    cpa, cpb = closest_points
+    return (cpa-cpb).length > cm.VTX_PRECISION
```
the downside to using the static `cm.VTX_PRECISION` is that it will behave differently if the input geometry is thousands of units in scale.. ( range 1000, 10 000 ).. isntead of the usual scene scale (0.0... 10.0 ) -- in EdgesIntersection node I think i made a manual 'scale' parameter accessible from the `draw_buttons_ext` area.

zeffii commented:
those things aside, this looks like a cool new tool!

zeffii commented:
i see (after more careful reading) you are using the Epsilon (mutable tolerance) good!

portnov commented:
Looks very intriguing.

portnov commented:
@vicdoval ,
1. If I'm not wrong, more correct name for this is "equidistant line".
2. About "SortbyConnexions". Am I getting it right that this is a topological sorting (https://en.wikipedia.org/wiki/Topological_sorting)? I was inspecting this question in https://github.com/nortikin/sverchok/issues/1990#issuecomment-354599848 (you may be interested in reading full discussion there). I'm not suggesting you to write this in C++ (huh), but you may consider rewriting my implementation of Tarjan's algorithm (https://github.com/portnov/python_cpp_excersises/blob/master/benchmark.py#L31) without use of recursion. I'm not ready yet to say anything about your sortVerticesByConnexions method, maybe it is using some well-known algorithm already? Then it worths to add a reference in comments.

vicdoval commented:
@zeffii Thanks for your kind word and excuse me for my innocence,  :) I applied the pep8 standards as you suggested

Following the avise of @portnov  I changed the name to avoid confusions, I thing "Contour 2D" fits pretty well its function. 
I changed the file name and class name to match the new name.

About the @portnov ideas on "SortingByConnexions":
I didn't even know  how this was called, and I don't have any base on graph theory. I did this function to solve my problem, but probably there are many better ways to do it. I checked your gist, but I would need to study it deeply to understand what you are doing.
Anyway thinking about the non-iteration thing I came up with a better solution that turned down the time needed from 2.5 seconds to 1.5 on 6000 vertices. :D 

So feel free to check the code and thanks for your feedback!

zeffii commented:
> @zeffii Thanks for your kind words and excuse me for my innocence, :)

i think you'll learn fast that pep8 (with a few ignore flags) results in more pleasant code formatting in general and promotes consistent Python style. It doesn't address algorithm implementation/difficulty, just makes code visually more accessible (mostly, not always).

I look forward to using the node, I can already imagine many typographic uses.

zeffii commented on nodes/modifier_make/contour2D.py:
```diff
@@ -456,7 +410,7 @@ class SvContourNode(bpy.types.Node, SverchCustomTreeNode):
                              update=updateNode)
 
-    rm_doubles = FloatProperty(name='R. Doubles', 
+    rm_doubles = FloatProperty(name='R. Doubles',
```
format them like this.
```python
    modeI = EnumProperty(
        name="modeI",
        description="Constant or weigted distance when multiple radius are given",
        items=modeItems, default="Constant",
        update=updateNode)

    listMatch = EnumProperty(
        name="listMatch",
        description="Behaviour on diffent list lengths",
        items=listMatchItems, default="Long Cycle",
        update=updateNode)

    rm_doubles = FloatProperty(
        name='R. Doubles',
        description="Remove Doubles Distance",
        min=0.0, default=0.0001,
        step=0.1, update=updateNode)
    
    epsilon = FloatProperty(
        name='Int. Tolerance',
        description="Intersection tolerance",
        min=1.0e-5, default=1.0e-5,
        step=0.02, update=updateNode)
    
    maskT = FloatProperty(
        name='Mask tolerance',
        description="Mask tolerance",
        min=-1.0, default=1.0e-5,
        step=0.02, update=updateNode)

    rad_ = FloatProperty(
        name='Distance', description='Contour distance',
        default=1.0, min=1.0e-5, update=updateNode)

    vert_ = IntProperty(
        name='N Vertices', description='Nº of Vertices per input vector',
        default=24, min=4, update=updateNode)
```

vicdoval commented on nodes/modifier_make/contour2D.py:
```diff
@@ -277,18 +275,20 @@ def CalcMidPoints(verts, edges):
+def maskEdges(edges, mask):
+    edges_out = []
+    for m, ed in zip(mask, edges):
+        if m == True:
```
I left it like that for some kind of code superstition :D

zeffii replied:
take a leap of faith for that kind of stuff.

zeffii commented:
for any functions that you take verbatim from other node files, you can import them instead. As long as those functions don't rely on some local storage to the source file. (here that's not the case)
```python
import sverchok.nodes.modifier_change.edges_intersect_mk2 as IX
from IX import remove_permutations_that_share_a_vertex
from IX import get_valid_permutations
```
etc...

vicdoval commented:
Great, but do i have to import de dependencies of that file? 
or by importing the file the dependencies will be automatically imported?

zeffii commented:
anything that those functions reference will be available to those functions _alone_ when imported, no namespace pollution.

zeffii commented:
take this short example of two files, one importing from the other.

#### file_a.py
```python
from sv_test import tk_dict_pusher, tk_dict_show

tk_dict_pusher("me")
tk_dict_pusher("too")
print(tk_dict_show())
```
##### sv_test.py
```python
some_dict = {}

def tk_dict_pusher(new_item):
    some_dict[new_item] = new_item

def tk_dict_show()
    return some_dict

```

zeffii commented:
if you can't get it working - that's OK. i'll consider turning those functions into a utility/module and guarantee that they are static

vicdoval commented:
I mange to import them 😄

zeffii commented on nodes/modifier_make/contour2D.py:
```diff
@@ -0,0 +1,599 @@
+
+
+def ptInTriang(p_test, p0, p1, p2):
+    # Function taken from Ramiro R.C https://stackoverflow.com/a/46409704
```
consider adding this to https://github.com/nortikin/sverchok/blob/master/utils/modules/geom_utils.py , and importing it.

zeffii commented on nodes/modifier_make/contour2D.py:
```diff
@@ -0,0 +1,599 @@
+                if A or B:
+                    d = 1
+                    break
+        if d > 0:
```
```diff
-        if d > 0:
-            mask.append(0)
-        else:
-            mask.append(1)
+        mask.append(0 if d > 0 else 1)
```

zeffii commented on nodes/modifier_make/contour2D.py:
```diff
@@ -0,0 +1,599 @@
+    edges_out = []
+    for m, ed in zip(mask, edges):
+        if m:
+                edges_out.append(ed)
```
indentation ++

zeffii commented on nodes/modifier_make/contour2D.py:
```diff
@@ -0,0 +1,599 @@
+
+            VertsAll = inputs['Verts_in'].sv_get(deepcopy=False)
+        else:
+            VertsAll = [[(0.0, 0.0, 0.0)]]
```
consider using the `default` parameter, which will be used in case `.is_linked` returns False
```diff
-        if inputs['Verts_in'].is_linked:
-
-            VertsAll = inputs['Verts_in'].sv_get(deepcopy=False)
-        else:
-            VertsAll = [[(0.0, 0.0, 0.0)]]

+       VertsAll = inputs['Verts_in'].sv_get(deepcopy=False, default=[[(0.0, 0.0, 0.0)]])
or (not sure about the extra wrapping)
+       VertsAll = inputs['Verts_in'].sv_get(deepcopy=False, default=[(0.0, 0.0, 0.0)])
```

zeffii commented on utils/modules/geom_utils.py:
```diff
@@ -74,3 +74,22 @@ def mean(verts):
     return dot_v3v3(normalized(n), normalized(normal_one)) < 0.0
+
+
+def ptInTriang(p_test, p0, p1, p2):
```
`def pt_in_triangle` this module has lowercase function names with words separated by `_` , as is common practice in Python code.

zeffii commented on nodes/modifier_make/contour2D.py:
```diff
@@ -0,0 +1,526 @@
+
+    def draw_buttons(self, context, layout):
+        layout.prop(self, "modeI", expand=True)
+        r = layout.row(align=True)
```
what's the assignment for?

zeffii commented on nodes/modifier_make/contour2D.py:
```diff
@@ -0,0 +1,526 @@
+        radiusAll = inputs['Radius'].sv_get(deepcopy=False, default=[[self.rad_]])
+        verticesAll = inputs['Nº Vertices'].sv_get(deepcopy=False, default=[[self.vert_]])
+        verticesAll = [list(map(lambda x: max(2, int(x)), Vertices)) for Vertices in verticesAll]
+        edgesAll = inputs['Edges_in'].sv_get(deepcopy=False, default=[[]])
```
: ) this is less chaotic

zeffii commented on nodes/modifier_make/contour2D.py:
```diff
@@ -0,0 +1,526 @@
+            listEdg.append((pointsL-1, 0))
+        return listEdg
+
+    def make_faces(self, Vertices):
```
skip the variable..
```python
def make_faces(self, Vertices):
    return [list(range(Vertices))]
```
oh..hehe i see you aren't using this function.

zeffii commented on nodes/modifier_make/contour2D.py:
```diff
@@ -0,0 +1,526 @@
+        edgesAll = inputs['Edges_in'].sv_get(deepcopy=False, default=[[]])
+
+        # outputs
+        if outputs['Vertices'].is_linked:
```
it's fine to check `*socket.is_linked` like you do here, but i recommend  using that to `return` early, so you can drop an indentation level in the rest of the block.
```python
if not outputs['Vertices'].is_linked:
    return

family = listMatcher([vertsAll, radiusAll, verticesAll, edgesAll], self.listMatch)
...
```

vicdoval replied:
cool alternative :)

zeffii commented on nodes/modifier_make/contour2D.py:
```diff
@@ -0,0 +1,526 @@
+                    totalPoints = 0
+                    for p in points:
+                        totalPoints += len(p)
+
```
`total_points = sum(len(p) for p in points)`  ..just a suggestion.

vicdoval commented:
Inside the Contour2D node:
the function Intersect Edges is an verbatim extract of IntersectEdges_mk2.py
the function MaskVertices  is an verbatim extract of vertices_mask.py

You think it would  be a good idea to extract the function in the original files and then importing it via  `import`?

portnov commented:
My personal "rule of thumb" for such questions is something like: if the block of code is used more than twice, OR if it is "complicated" in some sense — then it is usually better to factor it into separate method. There can be exceptions, mostly about code that uses a lot of variables from context (so that the extracted method would expect ten arguments or smth like that). But such code is anyway a candidate for refactoring :)

zeffii commented:
yeah i'm also leaning towards pushing more of these functions into shared modules
```
sverchok/utils/modules/<some_descriptive_name>_utils.py
```
if you need to customize the function you dupe it and append some form of indentifier like ..
```python
def original_func_name_<mod>(...):
    """
    this version of the function is used by node x, modified by y, for z reasons.
    """
```
just thinking out loud there.

vicdoval commented:
Images for docs
![contour2d_head](https://user-images.githubusercontent.com/10011941/34644861-37e4c430-f33f-11e7-92e0-d89080effc4b.png)
![multiple_distances_contour2d](https://user-images.githubusercontent.com/10011941/34644863-41eabfde-f33f-11e7-8ed6-6e8fa7a1e6df.png)
![multiple_objects_contour2d](https://user-images.githubusercontent.com/10011941/34644864-46463d24-f33f-11e7-80c1-bb0718d9966b.png)
![repeat_last_or_cycle_contour2d](https://user-images.githubusercontent.com/10011941/34644870-5935b1ee-f33f-11e7-99ba-0c536bf67f91.png)

zeffii commented on nodes/modifier_change/edges_intersect_mk2.py:
```diff
@@ -134,6 +134,41 @@ def unselect_nonintersecting(bm, d_edges, edge_indices):
+    return verts_out, edges_out
+
+
+def remove_doubles_with_edges(verts_in, edges_in, distance):
```
```diff
- def remove_doubles_with_edges(verts_in, edges_in, distance)
+ def remove_doubles_from_edgenet(verts_in, edges_in, distance)
```
`*_width_edges` had me confused, but `*_from_edgenet` might be more descriptive

zeffii commented:
about the mode / enum property

```python
modeItems = [
    ("Constant", "Constant", ""),
    ("Weighted", "Weighted", "")]

listMatchItems = [
    ("Long Repeat", "Long Repeat", ""),
    ("Long Cycle", "Long Cycle", "")]
```

This describes the enum list _minimally_, but it is less favourable to upgrades. This why most of our enumeration lists have explicit indices.
```
modeItems = [
    ("Constant", "Constant", "", 0),
    ("Weighted", "Weighted", "", 1)]
```
that way if you want to add enum members you can be sure that their indices won't be taken over by another enum member. (without indices you will encounter surprises if you add members to the middle of that list, _order matters when you don't supply indices_)

an option is to use a lambda/ function to generate the items
```python
lambda items: [(n, n, '', idx) for n, idx in items]

# leaving unused indices between 0 and 10 allows you to add items in between at a later stage.
nodeItems = items([["Constant", 0], ["Weighted", 10]])   
```
or.. if you want to have less `[...]` then
```python
def enum_items(*item_list):
    return [(n, n, '', idx) for n, idx in item_list]

nodeItems = enum_items(["Constant", 0], ["Weighted", 10])
```
------------

whatever you do, do add indices.

zeffii commented:
the last thing i want to say is that the process function is quite hefty,, can I encourage you to try and abstract parts of that into one or more function calls, and let the name of the function describe the inner working of that block you are abstracting away.

zeffii commented on nodes/modifier_make/contour2D.py:
```diff
@@ -0,0 +1,492 @@
+        edgesAll = inputs['Edges_in'].sv_get(deepcopy=False, default=[[]])
+
+        # outputs
+        if not outputs['Vertices'].is_linked:
```
this check can go directly at the top, stops all further processing if no outputs linked,

zeffii commented:
i think we can squash + merge this without too much further delay, if all nodes that have been edited exhibit no adverse effects.

zeffii commented:
@vicdoval  i like to get the process function down to almost nothing if possible, by abstracting 'actions' into functions which document themselves by having a reasonable function name. Here's step 1. notice that there's still a huge function which is definitely not easy to understand for an outsider/newcomer (me)

https://gist.github.com/zeffii/eaccc53befe30c54a76323dd74626cb1

**warning** i've not tested that - it might not work immediately. without persuading - i wish only to impress upon you the way we might abstract the more mundane stuff into functions.

step 2.  is going to break that `generate_outlines` function up, just enough to make it less imposing.

zeffii commented:
proof
```python
test_list = [[],[]]

def mutate_incoming(test_list, params):
    a, b = params
    list_a, list_b = test_list
    list_a.append(a)
    list_b.append(b)

a_list = [20,30,40]
b_list = ['huh', 'yes', 'ack']

for params in zip(a_list, b_list):
    mutate_incoming(test_list, params)

list_a, list_b = test_list
print(list_a, list_b)


>> [20, 30, 40] ['huh', 'yes', 'ack']
```

vicdoval commented:
When you talked about working with text I realized how slow was the node,  so i re-wrote everything upside down... at one point y stepped with the "intersect lines 2d" and that was an insane performance difference.
So I added the 2d routine to the intersect edges node...
I also made the changes  to empty de process function  :)

zeffii commented:
excellent @vicdoval ! did you see the changes @portnov made to the intersect_edges_mk2 node? (just because we don't see conflict warnings, doesn't mean...i can assume it's all going to be OK merging)

portnov commented:
https://gist.github.com/6d1433cf1825c447c5d6800fdbf53533

![curve_contour_1](https://user-images.githubusercontent.com/284644/35111694-a275fbe0-fc9d-11e7-96d8-ca85b7d9223b.png)

Interesting thing :)

zeffii commented:
@vicdoval  it's not clear to me why you remove the epsilon/min_dist ,  in the 2d version. Does the 2d version work for very large scale 10k+ bu mesh as well as much smaller than 1 bu scales?

zeffii commented:
am eager to merge this...

vicdoval commented:
It is a straight implementation of `mathutils.geometry.intersect_line_line_2d`, it is quite fast so there is no previous  edge removal.
The function returns `None ` in case of non intersecting pairs :)
Maybe more optimization could be made but  already hundred times faster than the 3d implementation and seems to work in all cases

zeffii commented:
cool. i'm going to merge unless you want to keep working on it before that..

vicdoval commented:
Last images for the docs
![contour9d](https://user-images.githubusercontent.com/10011941/35116834-027e2f8c-fc8d-11e7-9cff-35465e3e5e17.png)
![contour5d](https://user-images.githubusercontent.com/10011941/35116835-029ea8de-fc8d-11e7-9df0-f044677c059a.png)
![contour8d](https://user-images.githubusercontent.com/10011941/35116836-02b9cc36-fc8d-11e7-9526-259c18c8556f.png)

vicdoval commented:
I don't know if I did well the rst file, when i see it on github web looks like code, not like the other rst.
Any hint?

zeffii commented:
@vicdoval it might need to be on a sverchok branch directly to show the .rst as html.. it is weird i admit.

vicdoval commented:
OK, then I think I'm ready for merging  :D

zeffii commented:
@vicdoval congratulations and thank you for this node :)

vicdoval commented:
Thanks @zeffii , I really believe this is a great project and I'm glad i could place my grain of sand on it :)

zeffii commented:
i have a few edits to make to the commit.  (haha... it never ends)

vicdoval commented:
there is always some last edits... :D 

And the rst still does not render... maybe it needs some time :D

zeffii commented:
leave the .rst to me. i'll report back when it resolves

zeffii commented:
@vicdoval  under the word "Parameter" in the table you had a AWOL char `-` instead of `=` :)

zeffii commented:
sorry for bulldozing the other rst formatting.

vicdoval commented:
great that now it is working! 👍

zeffii commented:
![image](https://user-images.githubusercontent.com/619340/35121694-8b046d9e-fc9c-11e7-9e77-d6ccf819bcb8.png)

Durman commented:
@vicdoval I would like to ask is it possible to make the different modes working of this node such as make only inner contours or only outer contours or make contours to both directions how it works now? I already have tried to use this certainly useful node and I faced with problem to separate inner and outer contours.

Also it would not be bad to leave some comments about mode connect of vector sort. It seems does not work now.
https://gist.github.com/856dd6547cd0e7fed415d4ef33925c42
![2018-01-20_19-21-42](https://user-images.githubusercontent.com/28003269/35185091-02409b72-fe18-11e7-88ff-984307b0fce8.png)
![2018-01-20_19-29-36](https://user-images.githubusercontent.com/28003269/35185112-5badec3c-fe18-11e7-9a65-7abc09d7dde1.png)

vicdoval commented:
Thanks @Durman  I'm glad you find it useful :)
The `Contour 2D` node works with edges, so there is not a concept like inside or outside.
You can use `Separate Loose Parts` to get each contour
![contour2d_separate_contours](https://user-images.githubusercontent.com/10011941/35186547-3fea2800-fe16-11e7-8f70-dcadf181c239.png)

About the Vector Sort `Connect Mode`:
I've just made a pull request with the fix. 
It is meant to work with edges, specially with vertex with two connections. Check this gists https://gist.github.com/985bca8ca309c522b58108fd6368e08c
![vector_sort_connex](https://user-images.githubusercontent.com/10011941/35187803-3f88191c-fe2a-11e7-874b-da8cb4ec3751.png)

https://gist.github.com/7e6c9cf4494fa88439213ae179914f53
![vector_sort_connex1](https://user-images.githubusercontent.com/10011941/35187937-6352b0c0-fe2d-11e7-864c-a3893f94f6dc.png)

Durman commented:
@vicdoval Thanks, if outer contour always is first in list of contours then there is no problem. We used `topo sort for profile script` in `SNL` before, @zeffii  should be it deleted now?

zeffii commented:
@Durman  i'd consider making a folder in `node_scripts` for `old_scripts`, to keep the old stuff around as a reference. Not likely to delete the SNL script so soon, it's useful as a teaching tool.

stansu commented:
i found out that vector sort by connection won't work when edges is not closed.
https://gist.github.com/2afb7c38a6b00798fc051f449fc9f5fc

enzyme69 commented:
@stansu can try maybe a example Blend example and then we make new thread

stansu commented:
@enzyme69 we continue this topic in here #2100
